From a9518fe2c686de00320981567a2667de490de903 Mon Sep 17 00:00:00 2001
From: Brian Delwiche <delwiche@google.com>
Date: Tue, 6 May 2025 17:51:32 +0000
Subject: [PATCH] Check for existing discovery database in
 bta_hf_client_cb_init

Researcher finds an RCE related to HF client discovery database.

Changes against the related issue b/291281168 (guarded by flag below)
partially resolve this issue, but it is necessary to add a check in
another location.

Bug: 374746961
Test: m libbluetooth
Test: researcher POC
Ignore-AOSP-First: security
Tag: #security
Merged-In: Ie6329743c8acdc72154b9e3774ca53e840c85478
Change-Id: I69b2e80f0dd8682dcc8862b4b9aa7845c48c3bb8
---
 system/bta/hf_client/bta_hf_client_main.cc | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/system/bta/hf_client/bta_hf_client_main.cc b/system/bta/hf_client/bta_hf_client_main.cc
index c44f048481..8e003c8e23 100644
--- a/system/bta/hf_client/bta_hf_client_main.cc
+++ b/system/bta/hf_client/bta_hf_client_main.cc
@@ -309,6 +309,13 @@ void bta_hf_client_cb_init(tBTA_HF_CLIENT_CB* client_cb, uint16_t handle) {
   client_cb->enabled_hf_indicators.clear();
   client_cb->peer_hf_indicators.clear();
 
+  if (client_cb->p_disc_db) {
+    if (!get_legacy_stack_sdp_api()->service.SDP_CancelServiceSearch(client_cb->p_disc_db)) {
+      log::warn("Unable to cancel SDP service discovery peer:{}", client_cb->peer_addr);
+    }
+    osi_free_and_reset((void**)&client_cb->p_disc_db);
+  }
+
   // Memset the rest of the block
   // memset(client_cb, 0, sizeof(tBTA_HF_CLIENT_CB));
   *client_cb = {};
-- 
2.51.0.362.g3baa7cb742-goog

